tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs

```diff
@@ -0,0 +1,205 @@
+use crate::core_crypto::gpu::glwe_ciphertext_list::CudaGlweCiphertextList;
+use crate::core_crypto::gpu::CudaStreams;
+use crate::core_crypto::prelude::CiphertextModulusLog;
+use crate::integer::ciphertext::DataKind;
+use crate::integer::gpu::ciphertext::info::CudaBlockInfo;
+use crate::integer::gpu::ciphertext::{
+    CudaRadixCiphertext, CudaSignedRadixCiphertext, CudaUnsignedRadixCiphertext,
+};
+use crate::integer::gpu::list_compression::server_keys::{
+    CudaCompressionKey, CudaDecompressionKey,
+};
+
+pub struct CudaCompressedCiphertextList {
+    pub(crate) packed_list: (
+        CudaGlweCiphertextList<u64>,
+        Vec<CudaBlockInfo>,
+        CiphertextModulusLog,
+    ),
+    info: Vec<DataKind>,
+}
+impl CudaCompressedCiphertextList {
+    pub fn len(&self) -> usize {
+        self.info.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.info.len() == 0
+    }
+
+    pub fn get(
+        &self,
+        index: usize,
+        decomp_key: &CudaDecompressionKey,
+        streams: &CudaStreams,
+        gpu_index: u32,
+    ) -> CudaRadixCiphertext
+where {
+        let preceding_infos = self.info.get(..index).unwrap();
+        let current_info = self.info.get(index).copied().unwrap();
+
+        let start_block_index: usize = preceding_infos
+            .iter()
+            .copied()
+            .map(DataKind::num_blocks)
+            .sum();
+
+        let end_block_index = start_block_index + current_info.num_blocks() - 1;
+
+        decomp_key.unpack(
+            &self.packed_list,
+            start_block_index,
+            end_block_index,
+            streams,
+            gpu_index,
+        )
+    }
+}
+
+pub trait CudaCompressible {
+    fn compress_into(
+        self,
+        messages: &mut Vec<CudaRadixCiphertext>,
+        streams: &CudaStreams,
+    ) -> DataKind;
+}
+
+// Todo: Can we combine these two impl using CudaIntegerRadixCiphertext?
```

Comment:
I think you can remove this todo: the returned data types are different so you have to keep the two implementations of the function I think.